### ISSUE TITLE
chacha20/salsa20: performance optimizations (e.g. SIMD)

### ISSUE THREAD

**Issue by tarcieri** [RustCrypto/stream-ciphers#50]
There are two big optimizations we could do on both the `chacha20` and `salsa20` crates.

## Avoid recomputing initial state

- [ ] `chacha20` crate
- [ ] `salsa20` crate

[RFC 8439 Section 3](https://tools.ietf.org/html/rfc8439#section-3) describes caching the initial block state once computed as a performance optimization:

```
   Each block of ChaCha20 involves 16 move operations and one increment
   operation for loading the state, 80 each of XOR, addition and roll
   operations for the rounds, 16 more add operations and 16 XOR
   operations for protecting the plaintext.  Section 2.3 describes the
   ChaCha block function as "adding the original input words".  This
   implies that before starting the rounds on the ChaCha state, we copy
   it aside, only to add it in later.  This is correct, but we can save
   a few operations if we instead copy the state and do the work on the
   copy.  This way, for the next block you don't need to recreate the
   state, but only to increment the block counter.  This saves
   approximately 5.5% of the cycles.
```

## SIMD support

Both ChaCha20 and Salsa20 are amenable to SIMD optimizations. We should add SIMD optimizations on `x86`/`x86_64` at the very least.

- `x86`/`x86_64`
  - [ ] `chacha20`
  - [ ] `salsa20`
- Other CPU architectures?

**Comment by tarcieri** [comment]
Changed topic to `salsa20` as `chacha20` is now optimized on x86.

`chacha20` could still use e.g. NEON acceleration.

**Comment by paddor** [comment]
Long time no see! At least a decade later than you, I'm also expanding from Ruby to Rust lol.

I'm kinda working on SIMD accelerated salsa20 so that my OMQ.rs can match libzmq's CURVE perf. Right now it's still in a local crate called danza20 (I did chacha20 as well, failed to see the new release) but perf-wise it's looking very good. I've kinda ignored the whole constant time thing so far, given that it's a stream cipher that doesn't depend on input bits... but I should probably find a way to actually verify. I don't have experience in that yet.

Big fat warning: I'm still no actual "cryptographer" and I do be using AI.

Would you, like, categorically at all, consider a PR in this repo for that?

**Comment by tarcieri** [comment]
Yes, it'd be great if we had SIMD optimizations as the Salsa20 family is still heavily utilized in password hashing algorithms.

@paddor which CPU architectures are you optimizing for?

Regarding constant-time, it's still very much important but also you'd have to go out of your way to implement Salsa20 in a non-constant-time manner as it's easily amenable to constant-time implementations.

Anyway, yes, please open one or more PRs.

**Comment by paddor** [comment]
That's good to hear. I only have a 7-year old Intel CPU. No AVX-512. I used `fearless_simd` because it's amazing and should automagically do the right thing on ARM as well? For zrip it worked like a charm. I learned about dudect and verified CT. I'll extract it from my new unpublished `salsa20-simd` crate and work on opening a PR.
